### ISSUE TITLE
systemd support for wayland systems

### ISSUE THREAD

**Issue by Kommynct** [CharaChorder/nexus#148]
The only good way to log characters on linux wayland systems is by using systemd in the same way as kanata/keyd and others do.

This would also work for x11 so it's just a better way of doing things all around linux-side.

**Comment by Raymo111** [comment]
@Kommynct the shared upstream library is xslendix/libvinput that is used to do input capture, you should file this issue there instead.

**Comment by xslendix** [comment]
@Raymo111 it is on Nexus to create such systemd service. It makes literally no sense for a library to run as a daemon.

**Comment by Raymo111** [comment]
@xslendix they said "best way to log characters", I'm not sure what they want.

@Kommynct can you clarify what you want?

**Comment by xslendix** [comment]
> @xslendix they said "best way to log characters", I'm not sure what they want.

Yes which is literally the whole point of Nexus. You can have a systemd service that just executes Nexus. That's literally all there is to it. 

```ini
[Install]
WantedBy=graphical-session.target
```

**Comment by Kommynct** [comment]
Simple, nexus is meant to be used to log character entry to find lists of most commonly used words, as the note in the readme says, it lacks wayland compatibility, the only way to get full wayland compat working properly is a system service, I want this working on all wayland systems instead of just kde with a major security risk enabled

**Comment by Kommynct** [comment]
https://github.com/jtroo/kanata

https://github.com/rvaiya/keyd

^^these work independently of X and wayland, they allow remapping your keyboards by creating a virtual input device, something very similar could be used here to get rid of the dependency on X.

**Comment by xslendix** [comment]
@Raymo111 From what I understand, this would require reworking how the structure of the application works, having it run as some sort of server that continuously runs in the background. The GUI can stay there as it is currently, but instead of working with logic it interfaces with the server's API. For example you could have an HTTP server that has a /log/begin or /log/status or /log/words?page=1 etc

**Comment by Raymo111** [comment]
> @Raymo111 From what I understand, this would require reworking how the structure of the application works, having it run as some sort of server that continuously runs in the background. The GUI can stay there as it is currently, but instead of working with logic it interfaces with the server's API. For example you could have an HTTP server that has a /log/begin or /log/status or /log/words?page=1 etc

I don't think this has anything to do with the GUI, since there's a fully featured CLI. Didn't you tell me at one point that with the uinput rewrite, libvinput can be used in TTYs?

**Comment by xslendix** [comment]
> Didn't you tell me at one point that with the uinput rewrite, libvinput can be used in TTYs?

This is still true.

> I don't think this has anything to do with the GUI, since there's a fully featured CLI. 

It does, those need to be switched from calling the logger logic directly to using the API of a *new* Nexus daemon. You could have the server start and have the lifetime of the interface though and it would be the same as it is now. This ensures keys are captured if you want even at boot time, where the rest of services are started. I think this is what OP wants.

**Comment by Kommynct** [comment]
To be clear, I do want that, but I also want to ensure there's no xwayland dependency whatsoever, due to, again, the massive security implications.

**Comment by xslendix** [comment]
> I also want to ensure there's no xwayland dependency whatsoever.

xwayland is just a way to interface with X11 applications on Wayland compositors. `libvinput`, and Nexus *need* to support X11 as well, so support for it cannot be removed. If there's no x11 session, `libvinput` shrimply just uses the uinput kernel API.

**Comment by Kommynct** [comment]
there being no xwayland dependency doesn't mean remove support for x11, it just means on wayland, if there is no xwayland at all, it should still work.

However, i'm fairly certain they actually do not need to support x11, for the same reason keyd and kanata do not need to support x11, they use lower-level interception than either wayland or x11 and as a result do not depend on them at all. Again, doing things this way is more secure and will work on everything, even like, arcan.

**Comment by Raymo111** [comment]
> there being no xwayland dependency doesn't mean remove support for x11, it just means on wayland, if there is no xwayland at all, it should still work.

Currently on Wayland, nexus works out of the box, with no need for XWayland. I don't think anything else is needed.

**Comment by Kommynct** [comment]
![image](https://github.com/user-attachments/assets/09d516dd-9a41-4a53-b00c-9487d6800807)
this section of the readme should be updated, then
